File: app/src/main/java/com/example/bluetooth/MainActivity.java
package com.example.bluetooth;

import android.bluetooth.BluetoothDevice;
import android.os.Bundle;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ListView;

import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    private JDY08Manager manager;
    private ArrayAdapter<String> listAdapter;
    private final List<BluetoothDevice> devices = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        manager = new JDY08Manager(this);

        ListView listView = findViewById(R.id.list_devices);
        listAdapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1);
        listView.setAdapter(listAdapter);

        manager.startScan(device -> runOnUiThread(() -> {
            if (!devices.contains(device)) {
                devices.add(device);
                listAdapter.add(device.getName() + " - " + device.getAddress());
            }
        }));

        listView.setOnItemClickListener((parent, view, position, id) -> {
            manager.stopScan();
            BluetoothDevice device = devices.get(position);
            manager.connect(device);
        });

        Button onButton = findViewById(R.id.button_on);
        Button offButton = findViewById(R.id.button_off);

        onButton.setOnClickListener(v -> manager.sendCommand("1"));
        offButton.setOnClickListener(v -> manager.sendCommand("0"));
    }
}

File: app/src/main/java/com/example/bluetooth/MainActivity.kt
package com.example.bluetooth

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothSocket
import android.os.Bundle
import android.widget.Button
import androidx.appcompat.app.AppCompatActivity
import java.io.IOException
import java.util.UUID

class MainActivity : AppCompatActivity() {
    private val deviceAddress = "00:00:00:00:00:00" // TODO: set your module MAC address
    private val uuid: UUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB")
    private var bluetoothSocket: BluetoothSocket? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        findViewById<Button>(R.id.button_on).setOnClickListener {
            sendCommand("1")
        }
        findViewById<Button>(R.id.button_off).setOnClickListener {
            sendCommand("0")
        }
    }

    private fun sendCommand(cmd: String) {
        if (bluetoothSocket == null || !bluetoothSocket!!.isConnected) {
            connectSocket()
        }
        try {
            bluetoothSocket?.outputStream?.write(cmd.toByteArray())
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun connectSocket() {
        val adapter = BluetoothAdapter.getDefaultAdapter()
        val device: BluetoothDevice = adapter.getRemoteDevice(deviceAddress)
        bluetoothSocket = device.createRfcommSocketToServiceRecord(uuid)
        adapter.cancelDiscovery()
        try {
            bluetoothSocket?.connect()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}